src/mpid/common/thread/mpidu_thread_fallback.h

```diff
@@ -160,39 +174,80 @@ M*/
         }                                                               \
     } while (0)
 
+/* MPIDUI_THREAD_CS_YIELD in two parts, to allow code in between unlocked */
+#define MPIDUI_THREAD_CS_YIELD_BEGIN(mutex) \
+    do {                                                                \
+        if (MPIR_ThreadInfo.isThreaded) {                               \
+            /* always exit */                                           \
+            mutex.count--;                                              \
+            MPIR_Assert(mutex.count >= 0);                              \
+            mutex.owner = 0;                                            \
+            int err_ = 0;                                               \
+            MPL_DBG_MSG_P(MPIR_DBG_THREAD,VERBOSE,"MPIDU_Thread_mutex_unlock %p", &mutex); \
+            MPIDU_Thread_mutex_unlock(&mutex, &err_);                   \
+            MPIR_Assert(err_ == 0);                                     \
+        }                                                               \
+    } while (0)
```

Comment:
This code looks hacking and incorrect.  You are forcefully unlocking the mutex even when the mutex.count is not zero.  So, some other thread could get the mutex before this thread gets it again.

I don't like the general logic of this entire PR.  My recommendation is to not do this hack at all, and instead return a noncatastropic error up to the MPI-level function.  The MPI-level function can then decide to retry if it likes.

Reply:
> instead return a noncatastropic error up to the MPI-level function. The MPI-level function can then decide to retry if it likes.

I am not sure I understand what do you mean by `MPI-level` function. Do you mean ch4-layer? Either way, the upper-layer have no context on what the error mean or cause, how would it decide the "retry" logic? In addition, adding a retry logic at any layer will be as complicated and ugly, probably much more, as in the ofi-layer right now, right?

Reply:
Of course it's ugly.  But this hack it even more uglier. :-)

Conceptually, this is a transient error.  The network subsystem is out of buffer space.  The question really is how we handle this transient error.  What we want to do is to give up all the locks and then get back into the progress engine.  Your code for giving up all the locks is incorrect because it's blindly giving up the lock even when it's recursive.  By going all the way back up to the MPI layer (remember that that's the MPI layer for global lock), we know that the lock management code is correct.

An alternative approach is to not rely on inject calls always completing.  We can try to do `fi_inject`, and if it fails we fallback to `fi_write`.

Reply:
It is not just `fi_inject`, we use this macro, `MPIDI_OFI_CALL_RETRY` on nearly every libfabric calls, e.g. `fi_recvmsg`, `fi_send`, `fi_writemsg`...

Reply:
@wesbland mentioned that the reason of `FI_EAGAIN` ultimately is provider specific. We only can guess what is the cause.

Reply:
Why do we use it for nonblocking calls?  That doesn't make sense to me.

The reason for `FI_EAGAIN` is resource exhaustion, so the caller needs to try again.

Reply:
Nonblocking send/recv still assumes the operation will successfully enqueue the operation. Libfabric may return `FI_EAGAIN` when it fails at enqueue, presumably. Following are from doc:
```
-FI_EAGAIN : Indicates that the underlying provider currently lacks the resources needed to initiate the requested operation. The reasons for a provider returning FI_EAGAIN are varied. However, common reasons include insufficient internal buffering or full processing queues.

Insufficient internal buffering is often associated with operations that use FI_INJECT. In such cases, additional buffering may become available as posted operations complete.

Full processing queues may be a temporary state related to local processing (for example, a large message is being transferred), or may be the result of flow control. In the latter case, the queues may remain blocked until additional resources are made available at the remote side of the transfer.

In all cases, the operation may be retried after additional resources become available. It is strongly recommended that applications check for transmit and receive completions after receiving FI_EAGAIN as a return value, independent of the operation which failed. This is particularly important in cases where manual progress is employed, as acknowledgements or flow control messages may need to be processed in order to resume execution.
```

Reply:
Thanks.  If the provider runs out of queuing space, we should enqueue the operation in MPICH, or we can simply return the error back to the user.

Reply:
> ... we should enqueue the operation in MPICH
Then when do we retry the enqueued operation? Effectively it'll become an extended workq, right?

> we can simply return the error back to the user

This doesn't seem right. Why would user care about an transient state when the user code has nothing wrong?

Reply:
> Then when do we retry the enqueued operation? Effectively it'll become an extended workq, right?

Correct.  We'd try to make progress on the queued operations when we hit the progress engine again.  Just like workq or queued AMs in ch4.

> > we can simply return the error back to the user
> 
> This doesn't seem right. Why would user care about an transient state when the user code has nothing wrong?

I can understand your pushback against this, but if we run out of resources we do eventually need to return an error.  The only question is what we consider "running out of resources".

Reply:
> You are forcefully unlocking the mutex even when the mutex.count is not zero. So, some other thread could get the mutex before this thread gets it again.

You are right that the code isn't correct as now. Frankly, it is the whole recursive mutex code that is hacky -- but how about let's put the aesthetic judgment aside for now? :)

I also would like postpone the workq proposal. It is a lot of work and it just shifts a local issue into a general issue. Unless we return the `EAGAIN` error to user, who would have no idea of the cause, we still need poll and wait for the progress internally, inside ofi or inside ch4. Shall we agree on the postponing?

So right now I am facing the question of correctly yield the lock for full progress. I am tempted to spend the effort to ensure the recursive locking situation never arises (and fix it otherwise). Without recursive locking, yielding lock is a straight `EXIT/ENTER`. @pavan, what do you think?

Reply:
I agree that eliminating recursive locking is a good path forward.  But it might not be easy.

Reply:
> I agree that eliminating recursive locking is a good path forward. But it might not be easy.

Sounds good. In order to properly implement multi-vci locking, I'll need sorting out the double locking issues anyway.